ui: CSS modules for BarCharts components

koorosh opened the pull request:
Depends on: #47417
Depends on: https://github.com/cockroachdb/yarn-vendored/pull/20
Related to: #47527

Current draft is just example for possible styles isolation with CSS modules.
It is required to make components self-contained and easy for extraction.

Before all css files were loaded into
global scope even if file was imported in
some particular module.

Now it is possible to define styles as before
and use old styles without changes, and as a
module, to do this - files has to be named as
someName.module.styl
In webpack config, `style-loader` is defined
two times with different file name matchers
to be able define style loaders with and without
modules.

`barCharts` component is a first candidate to try out css modules.
- all styles related to `barCharts` are copied(!) to `barCharts.module.styl` file,
so in case another component somehow relies on styles defined for barCharts
- it won't be affected.

Storybook is extended with stories related to `barCharts`

koorosh committed:
ui: Add Statements page storybook

To test styles isolation for statements page
we need storybook which displays entire
Statements screen only.

To make it work, RouterProvider decorator
is added which connects router to dummy (empty)
store.

`statementsPage.fixture.ts` file contains snapshot
of required props for StatementsPage component

Release note: None

koorosh committed:
ui: Add css modules support for style-loader

Before all css files were loaded into
global scope even if file was imported in
some particular module.

Now it is possible to define styles as before
and use old styles without changes, and as a
module, to do this - files has to be named as
someName.module.styl
In webpack config, `style-loader` is defined
two times with different file name matchers
to be able define style loaders with and without
modules.

Release note: None

koorosh committed:
ui: CSS modules for barCharts component

- added storybook with barCharts to show
all possible types of barCharts
- Also typescript doesn't recognize `module.styl` files
as modules by default, so custom module declaration is added

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47484)
<!-- Reviewable:end -->

nathanstilwell commented on pkg/ui/src/views/statements/barCharts.module.styl:
```diff
@@ -0,0 +1,69 @@
+@require '~styl/base/palette.styl'
+@require '~src/components/core/index'
+
+.bar-chart
```
I dig this whole approach. And I'm excited to see us breaking components up this way since it will make it easier to reuse in the app. And if / when we want to use a component in another app, it will make it easier to move and publish.

One thing I look forward to with CSS Modules is losing the pesky BEM class naming since it no longer provides a function. I understand why you wouldn't do it here in the WIP, but I just wanted to call out that I think it's redundant to have the nesting under `.bar-chart`. So styles could become something like `${styles["bar"]}` instead of `${styles["bar-chart__bar"]}`

koorosh replied:
@nathanstilwell , I totally agree about BEM formatting, I preserved current names to:
- minimize code changes (easy to review and track lots of changes)
- to be able find code duplications which still remain in original stylings

nathanstilwell reviewed (commented):
I approve of this approach!

dhartunian reviewed (changes requested):
Very exciting @koorosh!! Just a few notes on the visuals in the storybook.

> Reviewed 5 of 8 files at r1, 2 of 8 files at r6, 1 of 1 files at r7, 5 of 5 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47484#-:-M5gvk5a6jFqqfWYodVt:b-wkxl7c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh and @nathanstilwell)

___
*[pkg/ui/src/views/statements/barCharts.stories.tsx, line 18 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/47484#-M5gvJGq1oSbr5QXg163:-M5gvJGq1oSbr5QXg164:bg6hkgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e953ca79bf7e3b37215e51a0a7dec4eca98e2713/pkg/ui/src/views/statements/barCharts.stories.tsx#L18)):*
> ```XML
> const { statements } = statementsPagePropsFixture;
> 
> storiesOf("BarCharts", module)
> ```

All the charts stop getting more narrow as I squish the browser window horizontally. Is there something preventing them from getting very small in the storybook? They have to get quite small to fit into the table columns.
___
*[pkg/ui/src/views/statements/barCharts.stories.tsx, line 27 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/47484#-M5gvEcG-LDwq7owKHIg:-M5gvEcHExLdJRDTFg7P:bpv85yy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e953ca79bf7e3b37215e51a0a7dec4eca98e2713/pkg/ui/src/views/statements/barCharts.stories.tsx#L27)):*
> ```XML
>     return chartFactory(statements[0]);
>   })
>   .add("retryBarChart", () => {
> ```

The retry bar chart should render in red. Also it shows a negative number in my storybook. Can we render with positive?
___
*[pkg/ui/src/views/statements/statementsPage.fixture.ts, line 23 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/47484#-M5gvbf_3LvhUN9drQWz:-M5gvbf_3LvhUN9drQX-:b-kno00e) ([raw file](https://github.com/cockroachdb/cockroach/blob/e953ca79bf7e3b37215e51a0a7dec4eca98e2713/pkg/ui/src/views/statements/statementsPage.fixture.ts#L23)):*
> ```XML
>   "count": Long.fromNumber(3),
>   "first_attempt_count": Long.fromNumber(100000),
>   "max_retries": Long.fromNumber(10),
> ```

Not sure why but I end up seeing this: 
![Screen Shot 2020-04-24 at 12.03.46 PM.png](https://files.reviewableusercontent.io/images/reviewable/986307/HcYOcxU18IPw8V_HdUQP85Wp/Screen%20Shot%202020-04-24%20at%2012.03.46%20PM.png)


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47484)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh and @nathanstilwell)

___
*[pkg/ui/src/views/statements/barCharts.tsx, line 141 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/47484#-M5gvzzf3UNFOKvmHQBR:-M5gvzzf3UNFOKvmHQBS:b-62vydd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e953ca79bf7e3b37215e51a0a7dec4eca98e2713/pkg/ui/src/views/statements/barCharts.tsx#L141)):*
> ```XML
>       };
> 
>       const className = classNames(styles["bar-chart"], styles[`bar-chart-${type}`], {
> ```

The `${type}` is what makes the bar chart red for retries so not sure what's missing...


<!-- Sent from Reviewable.io -->

koorosh commented:
@dhartunian, thanks for your suggestions, indeed there some gaps with BarCharts which have to be fixed.

> All the charts stop getting more narrow as I squish the browser window horizontally
Storybook had default `min-size` for the main panel. Removed this restriction so now Storybook more responsive.

Also to show bar charts within table columns - additional stories are added. Check out `BarCharts > within column (width 150px)`. It represents charts within the same constraints as in the Statements table.

> The retry bar chart should render in red. Also it shows a negative number in my storybook. Can we render with positive?

Fixed. The problem was with the wrong indentation for nested styles.

koorosh committed:
ui: Refactor classNames usage with bind for barCharts component

Previously, class names were constructed by simply accessing
style modules class names and assigning it to classes. It was
cumbersome and not readable at all.

To enhance this, `classnames/bind` alternate is used, which
allows simply put class names.

Release note: None

dhartunian reviewed (changes requested):
This all looks good from code perspective but causes a regression on the app right now. Could you take a look?

New:
![Screen Shot 2020-05-19 at 11.29.31 AM.png](https://files.reviewableusercontent.io/images/reviewable/986307/hM8X4cD-oWDualocyzW2Om5d/Screen%20Shot%202020-05-19%20at%2011.29.31%20AM.png)

Previous:
![Screen Shot 2020-05-19 at 11.29.32 AM.png](https://files.reviewableusercontent.io/images/reviewable/986307/3ugLBdySQdd7BcnrNlzOXi2Q/Screen%20Shot%202020-05-19%20at%2011.29.32%20AM.png)

> Reviewed 1 of 7 files at r9, 1 of 1 files at r10, 6 of 7 files at r11, 4 of 4 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47484#-:-M7hYUO-C3ZJv_S6iQXp:b-7848us)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh and @nathanstilwell)



<!-- Sent from Reviewable.io -->

koorosh commented:
> This all looks good from code perspective but causes a regression on the app right now. Could you take a look?
> 
> New:
> ![Screen Shot 2020-05-19 at 11.29.31 AM.png](https://camo.githubusercontent.com/582183c361a2d8fa61c7413c3b6f2dd171ab5202/68747470733a2f2f66696c65732e72657669657761626c6575736572636f6e74656e742e696f2f696d616765732f72657669657761626c652f3938363330372f684d38583463442d6f574475616c6f63797a57324f6d35642f53637265656e25323053686f74253230323032302d30352d3139253230617425323031312e32392e3331253230414d2e706e67)
> 
> Previous:
> ![Screen Shot 2020-05-19 at 11.29.32 AM.png](https://camo.githubusercontent.com/2d9db5128099400c50f5899d40dd320ada0b2f1a/68747470733a2f2f66696c65732e72657669657761626c6575736572636f6e74656e742e696f2f696d616765732f72657669657761626c652f3938363330372f3375674c426479535164643742636e724e6c7a4f586932512f53637265656e25323053686f74253230323032302d30352d3139253230617425323031312e32392e3332253230414d2e706e67)
> 
> > Reviewed 1 of 7 files at r9, 1 of 1 files at r10, 6 of 7 files at r11, 4 of 4 files at r12.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47484#-:-M7hYUO-C3ZJv_S6iQXp:b-7848us)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @koorosh and @nathanstilwell)

Regression with column width occurred because `barCharts` styles are overridden by outer component (`statementsTable`) and `statementsTable` is not yet work with CSS modules.
There is an example from `statements.styl` (refactored to CSS modules in another PR https://github.com/cockroachdb/cockroach/pull/48090)
``` css
.statements-table__col-rows, .statements-table__col-latency  // <-- CSS
  .bar-chart         // <-- trying to access class which is defined in CSS module
    min-width 150px
```

For me, it's a dilemma how to keep PR small enough with changes related directly to `barCharts` component and avoid such regressions.

Refactoring `statements.styl` to CSS modules is not an option because it will, in turn, require to refactor lots of modules that depend on it.

koorosh committed:
ui: Extend barCharts to accept external style props

Previously, Statements table altered styles of barCharts by
overriding them. It was necessary to set fixed width for barCharts
which were displayed in Statements table columns.
When `barCharts` were refactored to use CSS modules it became
impossible to override styles in such way (and it isn't react way).

Now, it is possible to extend `barCharts` styles by providing additional
 `options` param with `classess` field. It consumes `root` and `label` clasess, and then applies them inside of component.

Note, `statementsTable.module.styl` file has more styles than is used
right now, because this entire file was ported here from another PR and
this allows easily rebase it later.

Release note: None

koorosh commented:
@dhartunian , I managed to port changes related to custom styles for `barCharts` rendered in Statements table. It fixed the sizing of table columns with bar charts.
PS: It wasn't so big change as I thought before

dhartunian reviewed (approved):
Wonderful. Looks good to me. Let's ship it!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47484#-:-M7mkxtT2YJgpkXxlAh1:b-rjvgku)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @koorosh, and @nathanstilwell)



<!-- Sent from Reviewable.io -->

dhartunian commented:
bors r+

koorosh commented:
bors r+
